rust/psibase/src/flags.rs

```diff
@@ -0,0 +1,177 @@
+use crate::check;
+
+pub trait FlagsType {
+    type BitsetType: Copy
+        + std::ops::BitOr<Output = Self::BitsetType>
+        + std::ops::BitAnd<Output = Self::BitsetType>
+        + std::ops::Not<Output = Self::BitsetType>
+        + std::ops::Shl<u8, Output = Self::BitsetType>
+        + PartialEq
+        + From<u8>;
+
+    type JsonType: serde::Serialize + serde::de::DeserializeOwned + async_graphql::OutputType;
+
+    fn index(&self) -> u8;
+}
+
+/// # Flags
+///
+/// `Flags` represents a bitset structure that works with flag types implementing `FlagsType`.
+///
+/// Provides a convenient way to manage an integer that represents a set of boolean flags
+/// using strongly-typed indices.
+///
+/// # Type Parameter
+///
+/// * `F` - The flag type that implements `FlagsType`
+pub struct Flags<F: FlagsType> {
```

Comment:
I think Flags should be Packable and stored directly rather being cast to and from u8. It would also be nice if the representation is compatible with C++ `Bitset`.